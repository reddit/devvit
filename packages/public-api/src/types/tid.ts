import { assert } from '@devvit/shared-types/assert.js';

// CALMS!
export enum T_PREFIX {
  COMMENT = 't1_',
  ACCOUNT = 't2_',
  LINK = 't3_', // Same as POST
  MESSAGE = 't4_',
  SUBREDDIT = 't5_',
  AWARD = 't6_',
}

// string literal types
/** Commment thing ID. */
export type T1ID = `${T_PREFIX.COMMENT}${string}`;
/** Account thing ID. */
export type T2ID = `${T_PREFIX.ACCOUNT}${string}`;
/** Post (also called link) thing ID. */
export type T3ID = `${T_PREFIX.LINK}${string}`;
/** Message thing ID. */
export type T4ID = `${T_PREFIX.MESSAGE}${string}`;
/** Subreddit thing ID. */
export type T5ID = `${T_PREFIX.SUBREDDIT}${string}`;
/** Award thing ID. */
export type T6ID = `${T_PREFIX.AWARD}${string}`;

export type TID = T1ID | T2ID | T3ID | T4ID | T5ID | T6ID;

// type guards
export function isT1ID(id: string): id is T1ID {
  return id.startsWith(T_PREFIX.COMMENT);
}
export function isT2ID(id: string): id is T2ID {
  return id.startsWith(T_PREFIX.ACCOUNT);
}
export function isT3ID(id: string): id is T3ID {
  return id.startsWith(T_PREFIX.LINK);
}
export function isT4ID(id: string): id is T4ID {
  return id.startsWith(T_PREFIX.MESSAGE);
}
export function isT5ID(id: string): id is T5ID {
  return id.startsWith(T_PREFIX.SUBREDDIT);
}
export function isT6ID(id: string): id is T6ID {
  return id.startsWith(T_PREFIX.AWARD);
}

// assertion functions
export function assertT1ID(id: string): asserts id is T1ID {
  assert(isT1ID(id), `Expected comment id to start with ${T_PREFIX.COMMENT}, got ${id}}`);
}
export function assertT2ID(id: string): asserts id is T2ID {
  assert(isT2ID(id), `Expected account id to start with ${T_PREFIX.ACCOUNT}, got ${id}}`);
}
export function assertT3ID(id: string): asserts id is T3ID {
  assert(isT3ID(id), `Expected link id to start with ${T_PREFIX.LINK}, got ${id}}`);
}
export function assertT4ID(id: string): asserts id is T4ID {
  assert(isT4ID(id), `Expected message id to start with ${T_PREFIX.MESSAGE}, got ${id}}`);
}
export function assertT5ID(id: string): asserts id is T5ID {
  assert(isT5ID(id), `Expected subreddit id to start with ${T_PREFIX.SUBREDDIT}, got ${id}}`);
}
export function assertT6ID(id: string): asserts id is T6ID {
  assert(isT6ID(id), `Expected award id to start with ${T_PREFIX.AWARD}, got ${id}}`);
}

// factory functions
export function asT1ID(id: string): T1ID {
  assertT1ID(id);
  return id;
}
export function asT2ID(id: string): T2ID {
  assertT2ID(id);
  return id;
}
export function asT3ID(id: string): T3ID {
  assertT3ID(id);
  return id;
}
export function asT4ID(id: string): T4ID {
  assertT4ID(id);
  return id;
}
export function asT5ID(id: string): T5ID {
  assertT5ID(id);
  return id;
}
export function asT6ID(id: string): T6ID {
  assertT6ID(id);
  return id;
}

export function asTID<T extends TID>(id: string): T {
  if (isT1ID(id)) {
    return asT1ID(id) as T;
  }
  if (isT2ID(id)) {
    return asT2ID(id) as T;
  }
  if (isT3ID(id)) {
    return asT3ID(id) as T;
  }
  if (isT4ID(id)) {
    return asT4ID(id) as T;
  }
  if (isT5ID(id)) {
    return asT5ID(id) as T;
  }
  if (isT6ID(id)) {
    return asT6ID(id) as T;
  }
  throw new Error(
    `Expected thing id to start with ${Object.values(T_PREFIX).join(', ')} got ${id}}`
  );
}

// convenience functions
export function isCommentId(id: string): boolean {
  return isT1ID(id);
}
export function isAccountId(id: string): boolean {
  return isT2ID(id);
}
export function isLinkId(id: string): boolean {
  return isT3ID(id);
}
export function isMessageId(id: string): boolean {
  return isT4ID(id);
}
export function isSubredditId(id: string): boolean {
  return isT5ID(id);
}
export function isAwardId(id: string): boolean {
  return isT6ID(id);
}
