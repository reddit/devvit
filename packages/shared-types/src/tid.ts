import { assert } from './assert.js';

/**
 * - **C**omment (T1).
 * - **A**ccount (T2).
 * - **L**ink (T3).
 * - **M**essage (T4).
 * - **S**ubreddit (T5).
 * - **T**rophies (T6).
 */
export enum T_PREFIX {
  COMMENT = 't1_',
  ACCOUNT = 't2_',
  LINK = 't3_', // Same as POST
  MESSAGE = 't4_',
  SUBREDDIT = 't5_',
  AWARD = 't6_',
}

// string literal types
/** Commment thing ID. */
export type T1 = `${T_PREFIX.COMMENT}${string}`;
/** Account thing ID. */
export type T2 = `${T_PREFIX.ACCOUNT}${string}`;
/** Post (also called link) thing ID. */
export type T3 = `${T_PREFIX.LINK}${string}`;
/** Message thing ID. */
export type T4 = `${T_PREFIX.MESSAGE}${string}`;
/** Subreddit thing ID. */
export type T5 = `${T_PREFIX.SUBREDDIT}${string}`;
/** Award thing ID. */
export type T6 = `${T_PREFIX.AWARD}${string}`;

export type Tid = T1 | T2 | T3 | T4 | T5 | T6;

// type guards
export function isT1(id: string): id is T1 {
  return id.startsWith(T_PREFIX.COMMENT);
}
export function isT2(id: string): id is T2 {
  return id.startsWith(T_PREFIX.ACCOUNT);
}
export function isT3(id: string): id is T3 {
  return id.startsWith(T_PREFIX.LINK);
}
export function isT4(id: string): id is T4 {
  return id.startsWith(T_PREFIX.MESSAGE);
}
export function isT5(id: string): id is T5 {
  return id.startsWith(T_PREFIX.SUBREDDIT);
}
export function isT6(id: string): id is T6 {
  return id.startsWith(T_PREFIX.AWARD);
}

// assertion functions
export function assertT1(id: string): asserts id is T1 {
  assert(isT1(id), `Expected comment id to start with ${T_PREFIX.COMMENT}, got ${id}`);
}
export function assertT2(id: string): asserts id is T2 {
  assert(isT2(id), `Expected account id to start with ${T_PREFIX.ACCOUNT}, got ${id}`);
}
export function assertT3(id: string): asserts id is T3 {
  assert(isT3(id), `Expected link id to start with ${T_PREFIX.LINK}, got ${id}`);
}
export function assertT4(id: string): asserts id is T4 {
  assert(isT4(id), `Expected message id to start with ${T_PREFIX.MESSAGE}, got ${id}`);
}
export function assertT5(id: string): asserts id is T5 {
  assert(isT5(id), `Expected subreddit id to start with ${T_PREFIX.SUBREDDIT}, got ${id}`);
}
export function assertT6(id: string): asserts id is T6 {
  assert(isT6(id), `Expected award id to start with ${T_PREFIX.AWARD}, got ${id}`);
}

// factory functions
export function T1(id: string): T1 {
  assertT1(id);
  return id;
}
export function T2(id: string): T2 {
  assertT2(id);
  return id;
}
export function T3(id: string): T3 {
  assertT3(id);
  return id;
}
export function T4(id: string): T4 {
  assertT4(id);
  return id;
}
export function T5(id: string): T5 {
  assertT5(id);
  return id;
}
export function T6(id: string): T6 {
  assertT6(id);
  return id;
}

export function asTid<T extends Tid>(id: string): T {
  if (isT1(id)) {
    return T1(id) as T;
  }
  if (isT2(id)) {
    return T2(id) as T;
  }
  if (isT3(id)) {
    return T3(id) as T;
  }
  if (isT4(id)) {
    return T4(id) as T;
  }
  if (isT5(id)) {
    return T5(id) as T;
  }
  if (isT6(id)) {
    return T6(id) as T;
  }
  throw new Error(
    `Expected thing id to start with ${Object.values(T_PREFIX).join(', ')} got ${id}}`
  );
}
