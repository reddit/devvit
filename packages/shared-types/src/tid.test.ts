import { isT1, isT2, isT3, isT4, isT5, isT6, T_PREFIX, T1, T2, T3, T4, T5 } from './tid.js';

describe('Thing IDs', () => {
  it('never changes', () => {
    expect(T_PREFIX).toMatchSnapshot();
  });

  it('Tx_ Types are strict and correct', () => {
    const idSubstring = 'some_id_string';
    for (const prefix of Object.values(T_PREFIX)) {
      const idRaw = `${prefix}${idSubstring}`;
      switch (prefix) {
        case T_PREFIX.COMMENT:
          expect(isT1(idRaw)).toBe(true);
          expect(() => T2(idRaw)).toThrow();
          expect(() => T3(idRaw)).toThrow();
          expect(() => T4(idRaw)).toThrow();
          expect(() => T5(idRaw)).toThrow();
          break;
        case T_PREFIX.ACCOUNT:
          expect(isT2(idRaw)).toBe(true);
          expect(() => T1(idRaw)).toThrow();
          expect(() => T3(idRaw)).toThrow();
          expect(() => T4(idRaw)).toThrow();
          expect(() => T5(idRaw)).toThrow();
          break;
        case T_PREFIX.LINK:
          expect(isT3(idRaw)).toBe(true);
          expect(() => T1(idRaw)).toThrow();
          expect(() => T2(idRaw)).toThrow();
          expect(() => T4(idRaw)).toThrow();
          expect(() => T5(idRaw)).toThrow();
          break;
        case T_PREFIX.MESSAGE:
          expect(isT4(idRaw)).toBe(true);
          expect(() => T1(idRaw)).toThrow();
          expect(() => T2(idRaw)).toThrow();
          expect(() => T3(idRaw)).toThrow();
          expect(() => T5(idRaw)).toThrow();
          break;
        case T_PREFIX.SUBREDDIT:
          expect(isT5(idRaw)).toBe(true);
          expect(() => T1(idRaw)).toThrow();
          expect(() => T2(idRaw)).toThrow();
          expect(() => T3(idRaw)).toThrow();
          expect(() => T4(idRaw)).toThrow();
          break;
        case T_PREFIX.AWARD:
          expect(isT6(idRaw)).toBe(true);
          expect(() => T1(idRaw)).toThrow();
          expect(() => T2(idRaw)).toThrow();
          expect(() => T3(idRaw)).toThrow();
          expect(() => T4(idRaw)).toThrow();
          expect(() => T5(idRaw)).toThrow();
          break;
        default:
          throw new Error(`Unexpected prefix: ${prefix}`);
      }
    }
  });
});
