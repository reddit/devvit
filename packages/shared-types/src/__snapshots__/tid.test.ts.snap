// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Thing IDs > never changes 1`] = `
{
  "ACCOUNT": "t2_",
  "AWARD": "t6_",
  "COMMENT": "t1_",
  "LINK": "t3_",
  "MESSAGE": "t4_",
  "SUBREDDIT": "t5_",
}
`;
